### PR TITLE
Support custom transition duration for `DialogRoute`, `CupertinoDialogRoute` and show dialog methods.

### PR DESCRIPTION
**Pull request by nploi**
Currently we don't support custom transition duration for `DialogRoute`, `CupertinoDialogRoute` and show dialog methods , This PR will to support that.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Review by Piinks — COMMENTED** [review]
Hey @nploi thanks for the contribution! Can you add a test that verifies it works as expected? That will also ensure we do not undo this work in the future. Thanks!

**Comment by nploi** [comment]
> Hey @nploi thanks for the contribution! Can you add a test that verifies it works as expected? That will also ensure we do not undo this work in the future. Thanks!

Hi @Piinks, i added the tests for my changes. pls help to review. Thank you.

File: packages/flutter/lib/src/material/dialog.dart [hunk 1]
```diff
@@ -1428,6 +1428,7 @@ Future<T?> showDialog<T>({
   RouteSettings? routeSettings,
   Offset? anchorPoint,
   TraversalEdgeBehavior? traversalEdgeBehavior,
+  Duration transitionDuration = const Duration(milliseconds: 150),
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/dialog.dart:1431]: The `Material` widget has a `static const defaultSplashRadius` member:

https://github.com/flutter/flutter/blob/3bda455f4ef6fd626c66a986b69b512f412f7bab/packages/flutter/lib/src/material/material.dart#L439-L440

I think it'd be nice for `DialogRoute` and `CupertinoDialogRoute` to have something similar that stores the default duration—a single source of truth will make it easier to tweak in the future if needed.

File: packages/flutter/test/material/dialog_test.dart [hunk 2]
```diff
@@ -3002,15 +3093,26 @@ class _RestorableDialogTestWidget extends StatelessWidget {
 }
 
 class DialogObserver extends NavigatorObserver {
+  final List<ModalRoute<dynamic>> dialogRoutes = <ModalRoute<dynamic>>[];
   int dialogCount = 0;
```
Comment (nate-thegrate) [inline packages/flutter/test/material/dialog_test.dart:3097]: This change would help simplify the logic below.

```suggestion
  int get dialogCount => dialogRoutes.length;
```

(same applies to the Cupertino test as well)

**Review by nate-thegrate — COMMENTED** [review]
I left a couple of suggestions, but overall this is a fantastic improvement. Thanks!

File: packages/flutter/lib/src/material/dialog.dart (hunk 1 above)
Comment (nploi) [reply packages/flutter/lib/src/material/dialog.dart]: i added, thank you.

File: packages/flutter/test/material/dialog_test.dart (hunk 2 above)
Comment (nploi) [reply packages/flutter/test/material/dialog_test.dart]: i changed, thank you.

File: packages/flutter/test/material/dialog_test.dart [hunk 3]
```diff
@@ -2996,21 +3087,30 @@ class _RestorableDialogTestWidget extends StatelessWidget {
           },
           child: const Text('X'),
         ),
-      ),
+    ),
```
Comment (nate-thegrate) [inline packages/flutter/test/material/dialog_test.dart:3090]: Looks like an indent got removed randomly 🙃

```suggestion
      ),
```

**Review by nate-thegrate — APPROVED** [review]
LGTM, thanks for making this improvement!

File: packages/flutter/test/material/dialog_test.dart (hunk 3 above)
Comment (nploi) [reply packages/flutter/test/material/dialog_test.dart]: 😂 fixed.

File: packages/flutter/test/material/dialog_test.dart [hunk 4]
```diff
@@ -2829,6 +2873,53 @@ void main() {
     expect(find.text('Dialog2'), findsOneWidget);
   });
 
+
```
Comment (piedcipher) [inline packages/flutter/test/material/dialog_test.dart:2876]:
```suggestion
```

File: packages/flutter/test/material/dialog_test.dart [hunk 5]
```diff
@@ -2829,6 +2873,53 @@ void main() {
...
+
```
Comment (piedcipher) [inline packages/flutter/test/material/dialog_test.dart:2877]:
```suggestion
```

File: packages/flutter/test/material/dialog_test.dart [hunk 6]
```diff
@@ -2829,6 +2873,53 @@ void main() {
...
+    expect(rootObserver.dialogRoutes.length, equals(1));
+    final ModalRoute<dynamic> route = rootObserver.dialogRoutes.last;
+    expect(route is DialogRoute, true);
+    expect(route.barrierDismissible, isNotNull);
+    expect(route.barrierColor, isNotNull);
+    expect(route.transitionDuration, isNotNull);
+    expect(route.transitionDuration.inMilliseconds, 50);
+  });
+
+
```
Comment (piedcipher) [inline packages/flutter/test/material/dialog_test.dart:2922]:
```suggestion
```

**Review by piedcipher — COMMENTED** [review]
some smol clean-up

**Comment by nploi** [comment]
> some smol clean-up

thank for your review, i just fixed, pls help me check again 👍

**Review by dkwingsmt — COMMENTED** [review]
The changes look OK. The new argument needs documentation though.

**Review by piedcipher — APPROVED** [review]
LGTM

can we add set of tests for testing defaults too?

**Comment by piedcipher** [comment]
also one more point, wouldn't it be better to test whether transition happens post the duration instead of just checking the assignment?

**Comment by dkwingsmt** [comment]
> also one more point, wouldn't it be better to test whether transition happens post the duration instead of just checking the assignment?


I've also thought about this, but as long as we test `transitionDuration` works on `RawDialogRoute` then we're good to go.

And I've just checked that... it seems we have no such tests for `RawDialogRoute`.

So yeah, we need a test for `RawDialogRoute.transitionDuration`.

**Comment by nploi** [comment]
hi @dkwingsmt @piedcipher, Thank for your reviews. I just added docs, using nullable value and write a test for `RawDialogRoute.transitionDuration`, pls help to check it.

**Comment by nate-thegrate** [comment]
Looks like [**Linux Analyze**](https://ci.chromium.org/ui/p/flutter/builders/try/Linux%20analyze/86378/overview) is reporting a few trailing whitespace characters.

**Comment by nploi** [comment]
> Looks like [**Linux Analyze**](https://ci.chromium.org/ui/p/flutter/builders/try/Linux%20analyze/86378/overview) is reporting a few trailing whitespace characters.

fixed.

**Comment by nate-thegrate** [comment]
@nploi thanks for addressing all the feedback!

**Comment by TahaTesser** [comment]
@dkwingsmt @nate-thegrate 
We introduced `AnimationStyle` class to override animation durations and curves.

Instead of introducing new durations and curves parameters, we used this class to override `transitionDuration`
For instance here https://github.com/flutter/flutter/blob/2e221e7308ba58ef3b431e5b23c90927f6a0ffdf/packages/flutter/lib/src/material/popup_menu.dart#L909

**Comment by TahaTesser** [comment]
Just below this newly added `transitionDuration`, there are hard coded animation curves. If and when there is request to add support to customize these curves, we wouldn't want to add more new parameters.

image

Instead we could use `AnimationStyle` to override durations as well as curve, something like TextStyle for text

image

**Comment by TahaTesser** [comment]
I would recommend reverting this PR in favor of `AnimationStyle` for complete animation duration and curve customization.

**Comment by nate-thegrate** [comment]
Reason for revert: using `AnimationStyle` instead would allow for complete animation duration and curve customization.

**Comment by TahaTesser** [comment]
Since `AnimationStyle` relatively new,  for every widget I included a sample. We should also add sample for this change too.
Please look at an existing animation style  sample for default, custom, and no animation controls. You can copy this and replace the widget using new parameter.
https://api.flutter.dev/flutter/material/PopupMenuButton-class.html#material.PopupMenuButton.4

cc: @nate-thegrate

**Comment by nate-thegrate** [comment]
@TahaTesser thanks very much for this information!

I've created an issue to track progress on this feature: https://github.com/flutter/flutter/issues/154744

cc: @nploi

**Comment by dkwingsmt** [comment]
Thank you very much for this information! Learning something today!

**Comment by nploi** [comment]
> @TahaTesser thanks very much for this information!
> 
> I've created an issue to track progress on this feature: #154744
> 
> cc: @nploi

Thank you @nate-thegrate @TahaTesser, i will support `AnimationStyle` for show dialog methods.

can you assign me this https://github.com/flutter/flutter/issues/154744 @nate-thegrate ?

**Pull request merged**